noise.zig

```diff
@@ -0,0 +1,668 @@
+const std = @import("std");
+const crypto = std.crypto;
+const mem = std.mem;
+const testing = std.testing;
+const expect = testing.expect;
+
+const Ed25519 = crypto.sign.Ed25519;
+const X25519 = crypto.dh.X25519;
+const Blake2s256 = crypto.hash.blake2.Blake2s256;
+const ChaCha20Poly1305 = crypto.aead.chacha_poly.ChaCha20Poly1305;
+
+pub const HASH_LEN = 32;
+pub const KEY_LEN = 32;
+pub const BLOCK_LEN = 64;
+pub const MAC_LEN = 16;
+pub const MAX_MESSAGE = 65535;
+
+pub const EMPTY_KEY = [_]u8{0} ** 32;
+pub const EMPTY_HASH = [_]u8{0} ** 32;
+
+const Error = error{
+    InvalidMsgLen,
+};
+
+pub const NoiseSession = struct {
+    hs: HandshakeState,
+    h: [HASH_LEN]u8,
+    cs1: CipherState,
+    cs2: CipherState,
+    mc: u128,
+    i: bool,
+    is_transport: bool,
+
+    pub fn init_initiator(prologue: []const u8, s: Ed25519.KeyPair) NoiseSession {
+        return .{
+            .hs = HandshakeState.init(true, prologue, s, EMPTY_KEY),
+            .h = EMPTY_HASH,
+            .cs1 = CipherState.init(),
+            .cs2 = CipherState.init(),
+            .mc = 0,
+            .i = true,
+            .is_transport = false,
+        };
+    }
+
+    pub fn init_responder(prologue: []const u8, s: Ed25519.KeyPair) NoiseSession {
+        return .{
+            .hs = HandshakeState.init(false, prologue, s, EMPTY_KEY),
+            .h = EMPTY_HASH,
+            .cs1 = CipherState.init(),
+            .cs2 = CipherState.init(),
+            .mc = 0,
+            .i = false,
+            .is_transport = false,
+        };
+    }
+
+    pub fn send_msg(self: *NoiseSession, msg: []u8) !void {
+        if (msg.len < MAC_LEN or msg.len > MAX_MESSAGE)
+            return error.InvalidMsgLen;
+
+        if (self.mc == 0) {
+            try self.hs.write_msg_a(msg);
+        } else if (self.mc == 1) {
+            const tmp = try self.hs.write_msg_b(msg);
+            self.h = tmp.hash;
+            self.is_transport = true;
+            self.cs1 = tmp.cs1;
+            self.cs2 = tmp.cs2;
+            self.hs.clear();
+        } else if (self.i) {
+            self.cs1.write_msg_regular(msg);
+        } else {
+            self.cs2.write_msg_regular(msg);
+        }
+
+        self.mc += 1;
+    }
+
+    pub fn read_msg(self: *NoiseSession, msg: []u8) !void {
+        if (msg.len < MAC_LEN or msg.len > MAX_MESSAGE)
+            return error.InvalidMsgLen;
+
+        if (self.mc == 0) {
+            try self.hs.read_msg_a(msg);
+        } else if (self.mc == 1) {
+            const tmp = try self.hs.read_msg_b(msg);
+            self.h = tmp.hash;
+            self.is_transport = true;
+            self.cs1 = tmp.cs1;
+            self.cs2 = tmp.cs2;
+            self.hs.clear();
+        } else if (self.i) {
+            try self.cs2.read_msg_regular(msg);
+        } else {
+            try self.cs1.read_msg_regular(msg);
+        }
+
+        self.mc += 1;
+    }
+};
+
+pub const HandshakeState = struct {
```

Comment:
To make the state machine more safe, what about representing the `HandshakeState` has a union of structs? It'll also get rid of a lot of the null checks that happen down below.